[JENKINS-56243] Ensure user seed saved to session

jvz opened the pull request:
This fixes the remember me service to link the user seed to a
corresponding HTTP session. By doing so, this should significantly
reduce the number of calls to loadUserDetails().

See [JENKINS-56243](https://issues.jenkins-ci.org/browse/JENKINS-56243).

### Proposed changelog entries

* Fixed performance regression in the remember me service introduced by SECURITY-868.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @jeffret-b @Wadeck @daniel-beck @jenkinsci/code-reviewers 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

jvz committed:
[JENKINS-56243] Ensure user seed saved to session

This fixes the remember me service to link the user seed to a
corresponding HTTP session. By doing so, this should significantly
reduce the number of calls to loadUserDetails().

Signed-off-by: Matt Sicker <boards@gmail.com>

MarkEWaite reviewed (approved):
Looks good to me.  There are other locations in core that use a similar technique.

Is this a reasonable place to use a micro benchmark as described in recent blog posts on jenkins.io?

There are other uses of DISABLE_USER_SEED in Jenkins core.  One of those uses has a null check around the first reference to `user` after it has been assigned from the `User.get()` method.  I assume a null check is not needed for `user` here because it can be confidently predicted to never be null.

Wadeck commented:
@MarkEWaite Concerning the `User.get(Authentication auth);`, that method never returns null as long as you know that the authentication is not null and not anon, it will create the User if not present.

@jvz the test is nice to test that the user seed is present in the session, but it's just internal implementation. The expected behavior is that the session is "valid" and "usable" by the authentication filter. To test it, you can perhaps implement a mock SecurityRealm that provide a test-controlled userDetailsService.

jvz committed:
[JENKINS-56243] Improve tests

This validates that the performance regression itself - redundant calls
to loadUserByUsername() - is also fixed besides simply the
implementation details of the user seed and session.

Signed-off-by: Matt Sicker <boards@gmail.com>

Wadeck reviewed (approved):
Thanks for the test modification, it's a lot better now 👍

oleg-nenashev reviewed (approved):
LGTM. Will merge tomorrow if no negative feedback

jeffret-b reviewed (approved):
Looks great. This should make definite performance improvements.

daniel-beck reviewed (approved):
🎉
